laiaaa0/assignment1/q2.cpp

```diff
@@ -1,126 +1,127 @@
 #include <iostream>
+#include <stdexcept>
 using namespace std;
-
+/*
+Check for loops
+Free all elements when list is destroyed
+Unique pointers
+*/
 template <class T>
 class ListNode{
   private:
-    T value;
-    ListNode * next;
+    T node_value_;
+    ListNode * node_next_;
   public:
     ListNode(){
-
-        this->next = NULL;
-
+        this->node_next_ = NULL;
     }
     ListNode(T v, ListNode * nextNode){
-      this->value = v;
-      this->next = nextNode;
+      this->node_value_ = v;
+      this->node_next_ = nextNode;
     }
 
-    T getValue(){
-      return value;
+    const T & GetValue(){
+      return this->node_value_;
     }
-    ListNode * getNext (){
-      return next;
+    ListNode * GetNext (){
+      return this->node_next_;
     }
-    void setNext (ListNode *n){
-      this->next = n;
+    void SetNext (ListNode *n){
+      this->node_next_ = n;
+      /*
+TODO
```

Comment:
The safest way to do this is to make the node's next field private (and not provide a setter). Then mark the list class as a friend of node so it's the only one that can actually access and modify it.
Then of course you need to provide methods to create and modify lists.